src/internal/pjsdb/iterators.go

```diff
@@ -0,0 +1,104 @@
+package pjsdb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/jmoiron/sqlx"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+)
+
+//TODO(Fahad): Too much copy/paste from pfsdb, see if we can move the iterator code into another package which we can share
+// between this and other db packages.
+
+type OrderByColumn struct {
+	Column string
+	Order  sortOrder
+}
+
+type sortOrder string
+
+const (
+	SortOrderNone   = sortOrder("")
+	SortOrderAsc    = sortOrder("ASC")
+	SortOrderDesc   = sortOrder("DESC")
+	defaultPageSize = 100
+)
+
+type Iterable interface {
+	JobRow | queueRecord
+}
+
+type IteratorConfiguration struct {
+	StartPage uint64
+	PageSize  uint64          // defaults to 100 if not set
+	OrderBys  []OrderByColumn // defaults to id ASC if not set.
+}
+
+func (c *IteratorConfiguration) orderBy() string {
+	if c.OrderBys == nil {
+		c.OrderBys = []OrderByColumn{{Column: "id", Order: SortOrderAsc}}
+	}
+	values := make([]string, len(c.OrderBys))
+	for i, col := range c.OrderBys {
+		values[i] = fmt.Sprintf("%s %s", col.Column, col.Order)
+	}
+	return "\nORDER BY " + strings.Join(values, ", ")
+}
+
+type pageIterator[T Iterable] struct {
+	query                   string
+	values                  []any
+	limit, offset, maxPages uint64
+	page                    []T
+	pageIdx                 int
+	pagesSeen               int
+}
+
+// if maxPages == 0, then interpret as unlimited pages
+func newPageIterator[T Iterable](query string, values []any, startPage, pageSize, maxPages uint64) pageIterator[T] {
```

Comment:
Related to the iterator refactor you alluded to in a TODO.

Never merged this, but it goes from `stream.Iterator[[]T]` to `stream.Iterator[T]`
https://github.com/pachyderm/pachyderm/pull/8959

Shouldn't block this PR.